src/main/java/com/amazon/ion/impl/bin/IonRawBinaryWriter_1_1.kt

```diff
@@ -0,0 +1,898 @@
+                    // Clean up any unused space that was pre-allocated.
+                    buffer.shiftBytesLeft(currentContainer.length.toInt(), lengthPrefixPreallocation)
+                    val zeroLengthOpCode = OpCode.STRUCT_LENGTH_0
+                    buffer.writeUInt8At(currentContainer.position, zeroLengthOpCode + contentLength)
+                } else {
+                    thisContainerTotalLength += writeCurrentContainerLength(lengthPrefixPreallocation)
+                }
+            }
+            PREFIXED_STRUCT_FS -> {
+                thisContainerTotalLength += 1 + writeCurrentContainerLength(lengthPrefixPreallocation)
+            }
+            TAGLESS_EEXP -> {
+                // Nothing to do here because there's no opcode, length, or end delimiter.
+            }
+            EEXP -> {
+                // Add this to account for the opcode/address
+                thisContainerTotalLength += currentContainer.metadataOffset
+            }
+            PREFIXED_EEXP -> {
+                // NOTE: We could check if the length is 0 to see if we can go back and rewrite this as a
+                // non-length-prefixed e-exp, but we won't because that's easier done in the managed writer, which
+                // already has knowledge of the macro signature.
+
+                // Add to account for the opcode, address, and length prefix
+                thisContainerTotalLength += currentContainer.metadataOffset + writeCurrentContainerLength(lengthPrefixPreallocation)
+            }
+            DIRECTIVE -> {
+                thisContainerTotalLength += 2 // For the start and end delimiters
+                buffer.writeByte(OpCode.DELIMITED_CONTAINER_END)
+            }
+            else -> throw IonException("Nothing to step out of.")
+        }
+
+        val justExitedContainer = containerStack.pop()
+        val newCurrentContainer = containerStack.peek()
+        // Update the length of the new current container to include the length of the container that we just stepped out of.
+        newCurrentContainer.length += thisContainerTotalLength
+        newCurrentContainer.numChildren++
+        this.currentContainer = newCurrentContainer
+    }
+
+    /**
+     * Writes the length of the current container and returns the number of bytes needed to do so.
+     * Transparently handles PatchPoints as necessary.
+     *
+     * @param numPreAllocatedLengthPrefixBytes the number of bytes that were pre-allocated for the length prefix of the
+     *                                         current container.
+     */
+    private fun writeCurrentContainerLength(numPreAllocatedLengthPrefixBytes: Int, lengthToWrite: Long = currentContainer.length): Int {
+        val lengthPosition = currentContainer.position + currentContainer.metadataOffset
+        val lengthPrefixBytesRequired = flexUIntLength(lengthToWrite)
+        if (lengthPrefixBytesRequired == numPreAllocatedLengthPrefixBytes) {
```

Comment:
Should this be `<=` or are we using patch points when excess space was preallocated?